### PR TITLE
teams and high chain links

### PR DESCRIPTION
**Pull request by xgess**
When playing new links on a team's sigchain, add a couple fields to the TeamSigChainState so we always know which is the last high link. Then we can start signing over and sending up the most recently seen high link for all new links. Performance improvements here we come. 🔥

### CODE REVIEW

File: go/teams/chain_test.go [hunk 1]
```diff
@@ -59,6 +60,87 @@ func TestTeamSigChainParse(t *testing.T) {
...
+	require.NoError(t, err)
+	assertHighSeqForTeam(t, tc, teamID, 5)
+	assertHighSeqForTeam(t, tc, subteamID, 2)
+
+	// Deleting a subteam will not create a high link.
+	subteamName := teamName + "." + sub
+	err = Delete(ctx, tc.G, &teamsUI{}, subteamName)
+	require.NoError(t, err)
+	assertHighSeqForTeam(t, tc, teamID, 5)
+}
```
Comment (heronhaye) [inline go/teams/chain_test.go]: maybe a test for readding an admin after the admin resets? It should work because you're comparing `UserVersion`s though

**Comment by mlsteele** [comment]
Since the cache is persistent after this patch clients will be able to load things out of the cache that either do or don't have `lastHighSeqno` set in `TeamSigChainState`. Simplest thing would be to [bust the cache](https://github.com/keybase/client/blob/master/go/teams/storage.go#L92) but then everyone would have to reload their teams so.. nope. You could keep this in mind and any code written in the future should be aware that if `GetLatestHigh{Seqno,LinkID}()` returns 0 it should turn around and bust that team out of the cache and try again.

Also are we supposed to be storing LinkIDs or SigIDs? Whichever the hprevs are.

**Comment by maxtaco** [comment]
hprevs are linkIDs

**Comment by heronhaye** [comment]
For some reason it seems the `prev` hash is called a `LinkID` but I'm not sure why

File: protocol/avdl/keybase1/teams.avdl [hunk 2]
```diff
@@ -361,6 +361,10 @@ protocol teams {
     Seqno lastSeqno;
     LinkID lastLinkID;
 
+    // The last high link procesed
```
[the remark this answers is not in the document]
Comment (xgess) [reply protocol/avdl/keybase1/teams.avdl]: lol. i copied it from above. will fix everywhere.

**Comment by maxtaco** [comment]
linkID is the sha256 hash of the outer v2 link

**Comment by mlsteele** [comment]
This function gets called an unfortunate 3810 times for keybasefriends. `prevAdmins` eats 400ms and `wasHighLink` 200ms on my laptop when cold-loading keybasefriends.

Sorry, I had a hunch and [measured](https://github.com/keybase/client/commit/5149897c2584c526da77a4bc38b1157bd4000a89)

I suspect using `sanityCheckMembers` or its output `roleUpdates` and checking whether there's any admins or owners involved there would be faster. Checking for promotion is just checking whether there's the new owners or admin lists are populated. Checking for demotion though still involves fetching from the UserLog. But only for the number of removals when there are removals, vs now where it's once per admin/owner link.

**Comment by xgess** [comment]
@mlsteele i'll go down this path if you can give me another performance test in a little bit. one quick thought though: changing a user from owner to admin (or vice versa) wouldn't be a high link, so I'm going to need to check that user's previous role anyway. :/ splatting the logic into the case statement instead of doing it all at once should dramatically reduce the number of times it runs. The big tradeoff will be obviousness of the code, but I'll see if some comments and descriptive variable names can't mitigate that. we might also be able to accomplish the same performance improvement with a couple extra `if` statements. i can explore both.

**Comment by mlsteele** [comment]
Yep I can do that or share some setup.

It's true that a link the changes an owner to an admin is not technically in the minimal set of high links. I think it would be ideal to [match the server](https://github.com/keybase/keybase/blob/master/helpers/sig.iced#L1115). Which considers such links high. I'm hoping that the server can check the client by asserting that they exactly agree on the 'observed high link set'. Idk whether we should tighten the server or use that same check here on the client. @heronhaye 

It seems like it will be pretty clear but easier to forget if another link type is added in the future that admins or owners. I think the `if` sprinkling alternative would result in sprawling control flow.

**Comment by mlsteele** [comment]
Total time when cold-loading keybasefriends. Solid 1000x improvement.
```
chain.isHighLink [time=21.919µs]
chain.roleUpdateChangedHighSet [time=654.135µs]
```

**Pull request merged**